src/Pdo/Query/BuildInsertSqlCapableTrait.php

```diff
@@ -0,0 +1,179 @@
+<?php
+
+namespace RebelCode\Storage\Resource\Pdo\Query;
+
+use Dhii\Util\String\StringableInterface;
+use Dhii\Util\String\StringableInterface as Stringable;
+use InvalidArgumentException;
+
+/**
+ * Common functionality for objects that can built INSERT SQL queries.
+ *
+ * @since [*next-version*]
+ */
+trait BuildInsertSqlCapableTrait
+{
+    /**
+     * Builds a INSERT SQL query.
+     *
+     * @since [*next-version*]
+     *
+     * @param string|Stringable     $table        The name of the table to insert into.
+     * @param string[]|Stringable[] $columns      A list of columns names. The order is preserved in the built query.
+     * @param array                 $rowSet       A list containing record data maps, mapping column names to row
+     *                                            values.
+     * @param array                 $columnMap    Optional mapping of field names to column names.
+     * @param array                 $valueHashMap Optional map of value names and their hashes.
+     *
+     * @return string The built INSERT query.
+     */
+    protected function _buildInsertSql(
+        $table,
+        array $columns,
+        array $rowSet,
+        array $columnMap = [],
+        array $valueHashMap = []
+    ) {
+        $tableName = $this->_escapeSqlReference($table);
+        $columnsList = $this->_escapeSqlReferenceArray($columns);
+        $values = $this->_buildSqlValuesList($columns, $rowSet, $columnMap, $valueHashMap);
+
+        $query = sprintf(
+            'INSERT INTO %1$s (%2$s) %3$s',
+            $tableName,
+            $columnsList,
+            $values
+        );
+
+        return sprintf('%s;', trim($query));
+    }
+
+    /**
+     * Builds the VALUES portion of an INSERT SQL query.
+     *
+     * @since [*next-version*]
+     *
+     * @param string[]|Stringable[] $columns      A list of columns names. The order is preserved in the built query.
+     * @param array                 $rowSet       A list containing record data maps, mapping column names to row
+     *                                            values.
+     * @param array                 $columnMap    Optional mapping of field names to column names.
+     * @param array                 $valueHashMap Optional map of value names and their hashes.
+     *
+     * @return string The built VALUES list or an empty string if the row set has no entries.
+     */
+    protected function _buildSqlValuesList(
+        array $columns,
+        array $rowSet,
+        array $columnMap = [],
+        array $valueHashMap = []
+    ) {
+        if (count($rowSet) === 0) {
+            return '';
+        }
+
+        $values = [];
+
+        foreach ($rowSet as $_rowData) {
+            $values[] = $this->_buildSqlRowValues($columns, $_rowData, $columnMap, $valueHashMap);
+        }
+
+        return sprintf('VALUES %s', implode(', ', $values));
```

Comment:
What do you think about leaving the `VALUES` word itself up to the consumer of this method?